### ISSUE TITLE
Positional encoding in DINOv2: Case without interpolation

### ISSUE THREAD

**Issue by v-espitalier** [huggingface/candle#2286]
Hi,

I have been reviewing DINOv2 Candle code and I noticed most likely a bug (unless I misunderstood the code).

As far as I understand, the function **interpolate_pos_encoding()** is used to interpolate the (transformer-wise absolute) positional encoding before being added to the transformer tokens 'xs':
```
fn prepare_tokens_with_mask(&self, xs: &Tensor) -> Result<Tensor> {
        let (_b, _nc, w, h) = xs.dims4()?;
        let xs = self.patch_embed.forward(xs)?;
        let xs = Tensor::cat(&[&self.cls_token, &xs], 1)?;
        &xs + &self.interpolate_pos_encoding(&xs, w, h)?
    }
```

But the interpolation function seems buggy in the case where there is no interpolation, and the "shortcut" is used.
See the beginning of the function:
```
 fn interpolate_pos_encoding(&self, xs: &Tensor, w: usize, h: usize) -> Result<Tensor> {
        let npatch = xs.dim(1)? - 1;
        let n = self.pos_embed.dim(1)? - 1;
        let sqrt_n = (n as f64).sqrt();
        if npatch == n && w == h {
            return Ok(xs.clone());
        }
        let class_pos_embed = self.pos_embed.i((.., ..1))?;
        let patch_pos_embed = self.pos_embed.i((.., 1..))?;
```

I guess inside the shortcut, we should return this instead:
```
        if npatch == n && w == h {
            return Ok(self.pos_embed)
        }
```
Otherwise, we are doubling the values of the tokens xs ?!

Thanks for your answer,

**Comment by zlogic** [comment]
Indeed, I was trying to use the DINOv2 model to make a Rust port of Apple Depth Pro, and it produced very poor results. Changing this line fixed my issue.

If rescaling/interpolation is needed, https://github.com/huggingface/candle/blob/17cbbe4286f25934197db79a244fd0694259c899/candle-transformers/src/models/dinov2.rs#L291 returns data from `pos_embed`, so I'd expect the other path to returns `pos_embed` as well.

The Python (timm) original version returns `pos_embed`, not `xs`: https://github.com/huggingface/transformers/blob/f19135afc77053834f1b0cdf46d9a6bf7faf7cc3/src/transformers/models/vit/modeling_vit.py#L92

**Comment by pcuenca** [comment]
Fixed in #3172, thanks @v-espitalier and @zlogic 🙌
